Fix mishandling of some class one-liners

mwoehlke opened the pull request:
Teach `nl_before_opening_brace_func_class_def` to respect `nl_class_leave_one_liners`.

Also, tweak `mark_cpp_constructor` to also set the parent type on any qualifiers ('`virtual`', '`constexpr`') preceding a ctor/dtor prototype or definition. This fixes `nl_class_leave_one_liner_groups` not being applied to constexpr constructors.
    
Fixes #2907.

mwoehlke commented:
Well, this is an interesting failure (30117):
```
class Capteur_CO2 : public Capteur {
    public:
        Capteur_CO2() :
            un_membre_en_plus ( 0 ) {}
};
```

...seems that's considered a one-liner? That... doesn't seem right?

gmaurel reviewed (commented):
Some problems at 31-test

mwoehlke commented:
> Some problems at 31-test

...as already noted in [my prior comment](https://github.com/uncrustify/uncrustify/pull/2908#issuecomment-689919272) (and in somewhat more detail [here](https://github.com/uncrustify/uncrustify/issues/2907#issuecomment-689920718)). I'm not sure what's the correct behavior here; I'd appreciate your thoughts?

guy-maurel commented:
I have made a test.
Merge your changes, updated the p.txt, 31.txt, 66.txt
I now get:
```.txt
Uncrustify_d-0.71.0-157-a784082a-dirty
Parsing: mwoehlke.cpp as language CPP
```
```.cpp
template< typename Enum > class Flags
{
public:
constexpr Flags() : value{ 0 } {}
constexpr Flags( Enum f ) : value( static_cast< value_t >( f ) ) {}
constexpr Flags( Flags const& ) = default;

}

```

Is it right?

mwoehlke commented:
Arf, I missed the CLI / sources-format issues. Those are probably easy; I'll try to take a look later. The one that's vexing me is cpp:30117. I could bash that into passing in a couple ways, but what is the *correct* behavior for that? Is the change in behavior there desired or not?

mwoehlke commented:
I've updated the PR so that this:
```c++
  class_ctor() :
    member{...} {}
```
...is **not** considered a one-liner for `nl_before_opening_brace_func_class_def`. (But I didn't change `one_liner_nl_ok`.)

Please consider if this seems like a reasonable approach.

I also fixed the CLI tests and formatting; hopefully everything is passing now.

mwoehlke commented:
> Please consider if this seems like a reasonable approach.

In fact, I am inclined (partly, after noticing that's the way my own code is formatted :stuck_out_tongue:) to *not* take that approach, and to indeed treat that as a one-liner. That would imply changing the expected output of cpp:30117, however.

mwoehlke commented:
> In fact, I am inclined to *not* take that approach, and to indeed treat that as a one-liner. That would imply changing the expected output of cpp:30117, however.

I've now pushed that version instead. Please take note of the changes to cpp:30117 and consider whether or not they seem reasonable.
